Plugin~/WebRTCPlugin/AudioTrackSinkAdapter.cpp

```diff
@@ -1,56 +1,82 @@
 #include "pch.h"
 #include "AudioTrackSinkAdapter.h"
+#include "common_audio/include/audio_util.h"
 
 namespace unity
 {
 namespace webrtc
 {
-    AudioTrackSinkAdapter::AudioTrackSinkAdapter(DelegateAudioReceive callback)
-        : _callback(callback)
+    AudioTrackSinkAdapter::AudioTrackSinkAdapter()
+        : m_ringBuffer(nullptr), m_bufferIn(nullptr), m_bufferInLen(0)
     {
     }
 
+    AudioTrackSinkAdapter::~AudioTrackSinkAdapter()
+    {
+        if (m_ringBuffer)
+        {
+            WebRtc_FreeBuffer(m_ringBuffer);
+            m_ringBuffer = nullptr;
+        }
+        if (m_bufferIn)
+        {
+            free(m_bufferIn);
+            m_bufferIn = nullptr;
+            m_bufferInLen = 0;
+        }
+    }
+
     void AudioTrackSinkAdapter::OnData(
         const void* audio_data, int bits_per_sample, int sample_rate,
         size_t number_of_channels, size_t number_of_frames)
     {
-        if (_callback == nullptr) {
-            return;
+        if (m_ringBuffer) {
+            std::lock_guard<std::mutex> lock(m_mutex);
+            webrtc::voe::RemixAndResample((const int16_t*) audio_data, number_of_frames, number_of_channels, sample_rate, &m_resampler, &m_frame);
+            WebRtc_WriteBuffer(m_ringBuffer, m_frame.data(), m_frame.num_channels() * m_frame.samples_per_channel());
         }
+    }
 
-        switch (sample_rate) {
-        case 44100:
-            break;
-        case 48000:
-            break;
-        default:
-            RTC_LOG(LS_WARNING) << "Unsupported sampling rate: " << sample_rate;
-            return;
-        }
+    void AudioTrackSinkAdapter::ProcessAudio(float* data, int32 size, int channels, int sampleRate)
+    {
+        int samplesLeft = size / channels;
+        int sampleIdx = 0;
 
-        if (bits_per_sample != 16) {
-            RTC_LOG(LS_WARNING) << "Unsupported bits/sample: " << bits_per_sample;
+        if (sampleRate == 0)
             return;
-        }
-
-        const size_t size = number_of_channels * number_of_frames;
-        const int16_t* data = static_cast<const int16_t*>(audio_data);
-        const float_t INVERSE = 1.0 / SHRT_MAX;
 
-        std::vector<float_t> _converted_data(size);
+        memset(data, 0, sizeof(float) * samplesLeft * channels);
```

Comment:
there shouldn't be a need for the memset but was for clarity at some point. It's unlikely that the while loop exits when there's not enough data, but could always do memset for the unfilled part